hail/src/main/scala/is/hail/io/fs/GoogleStorageFS.scala

```diff
@@ -51,9 +54,77 @@ object GoogleStorageFileStatus {
   }
 }
 
-class GoogleStorageFS(val serviceAccountKey: Option[String] = None) extends FS {
+object RequesterPaysConfiguration {
+  def fromFlags(requesterPaysProject: String, requesterPaysBuckets: String): Option[RequesterPaysConfiguration] = {
+    if (requesterPaysProject == null) {
+      if (requesterPaysBuckets == null) {
+        None
+      } else {
+        fatal(s"Expected requester_pays_buckets flag to be unset when requester_pays_project is unset, but instead found: $requesterPaysBuckets")
+      }
+    } else {
+      val buckets = if (requesterPaysBuckets == null) {
+        None
+      } else {
+        Some(requesterPaysBuckets.split(",").toSet)
+      }
+      Some(RequesterPaysConfiguration(requesterPaysProject, buckets))
+    }
+  }
+}
+
+case class RequesterPaysConfiguration(
+  val project: String,
+  val buckets: Option[Set[String]] = None
+) extends Serializable
+
+class GoogleStorageFS(
+  private[this] val serviceAccountKey: Option[String] = None,
+  private[this] var requesterPaysConfiguration: Option[RequesterPaysConfiguration] = None
+) extends FS {
   import GoogleStorageFS._
 
+  def getConfiguration(): Option[RequesterPaysConfiguration] = {
+    requesterPaysConfiguration
+  }
+
+  def setConfiguration(config: Any): Unit = {
+    requesterPaysConfiguration = config.asInstanceOf[Option[RequesterPaysConfiguration]]
+  }
+
+  private[this] def requesterPaysOptions[T](bucket: String, makeUserProjectOption: String => T): Seq[T] = {
+    requesterPaysConfiguration match {
+      case None =>
+        Seq()
+      case Some(RequesterPaysConfiguration(project, None)) =>
+        Seq(makeUserProjectOption(project))
+      case Some(RequesterPaysConfiguration(project, Some(buckets))) =>
+        if (buckets.contains(bucket)) {
+          Seq(makeUserProjectOption(project))
+        } else {
+          Seq()
+        }
+    }
+  }
+
+  private[this] def handleRequesterPays[T, U](
+    makeRequest: Seq[U] => T,
+    makeUserProjectOption: String => U,
+    bucket: String
+  ): T = {
+    try {
+      makeRequest(Seq())
+    } catch {
+      case exc: StorageException =>
```

Comment:
This is pretty gross. Why don't we always include the RP logic?

Reply:
If you supply a userProject it will be billed for the charges, even if the bucket is not requester pays. Consider, for example, a non-requester-pays bucket B1 owned by project P1. If a user (with read access to B1) issues a request to get the object identified by gs://B1/O1 with no userProject, then the class A operation will be billed to P1. If, instead, the user issues a request with userProject set to P2, the class A operation will be billed to P2.

Google does not provide a method for a non-owning-user to determine the requester pays status of a bucket. Ergo, the only way to know that a bucket is requires the userProject argument is to make the request, catch the error, and check the reason or the message. Apparently the API can alternately produce a message with a "reason" set to "userProjectMissing" or produce a message with an uninformative reason but include something about requester pays in the message.

It's a terrible API.

Reply:
I suppose an alternative we could pursue is this: the user must specify, a priori, all requester pays buckets they intend to use and the project to which to bill that usage. e.g.:

```python3
hl.init(requester_pays_config=(['bucket1', 'bucket2', 'bucket3', ...], 'my-project'))
```

We can then provide the userProject unconditionally on any request to bucket1, bucket2, bucket3, etc.